sw/device/lib/dif/README.md

```diff
@@ -50,6 +51,30 @@ there are some relaxations of these rules for them described at the end.
 * DIFs must use enum return codes rather than booleans for reporting errors. If
   a DIF can either error or instead produce a value, it should return an enum
   return code, and use an out-parameter for returning the produced value.
+  * DIF libraries must have a general return code enum called
+    `dif_<ip>_result_t`. It must define the following constants (in this order):
+    * `kDif<ip>Ok`, with value 0, to denote the call succeeded.
+    * `kDif<ip>Error`, to denote a non-specific error happened during the call.
+      This is for the `default:` case of enum switches (as noted below), and for
+      assertion errors (usually where the function has already caused side
+      effects so `kDif<ip>BadArg` cannot be used).
+    * `kDif<ip>BadArg`, to denote that the caller supplied incorrect arguments.
+      This value must only be returned before the function does anything that
+      causes side effects.
+  * Other return code enums can be defined for functions that fail in more
+    specific ways, and these other return code enums can be shared between
+    multiple functions that fail in the same way.
+    These more specific return codes must be named
+    `dif_<ip>_<operation>_result_t`, their constants should be prefixed with
+    `kDif<ip><operation>`, and the first three constants should be the same as
+    the three constants in `dif_<ip>_result_t`.
+    The additional, specific return code constants must all be defined after
+    the three general constants.
+  * DIFs must document the meaning of all error codes, including to what extent
+    the error is recoverable, or to what extent the operation is retryable.
+    `kDif<ip>Error` is not recoverable. `kDif<ip>BadArg` is designed to be
```

Comment:
Is `kDif<ip>BadArg` really recoverable? To me this implies that the caller unknowingly supplied an invalid argument, and therefore represents a bug in the implementation of the driver supplying the arguments to the respective DIF. It seems to me that trying to implement recovery for a bug that the driver itself caused only increases complexity, and I can't imagine a test case for this code path that wouldn't catch the bug causing the `BadArg` in the first place.

Reply:
I agree, not sure if `kDif<ip>BadArg` is recoverable. If you pass it `null`, you presumably didn't do it on purpose, and there is not much you can do about it.

However, if you add meaningful error codes to functions like `dif_uart_configure`, you possibly can recover from things like `kDifUartBadNco` **(just an example)**.

Reply:
I am guessing unrecoverable errors will result in some drastic action. If so, it is reasonable to classify `kDif<ip>BadArg` as recoverable, especially in a shared context. Consider the following:
- There are multiple apps, using the same device.
- They share their buffers with the device DIF library (probably via the device capsule in tock, not clear if the capsule will perform the same checks as DIF libraries do, I am assuming it doesn't but even if it does the same discussion would apply to it too)
- BadApp has a bug and passes a `NULL` pointer.

Even if BadApp cannot meaningfully retry the call, we may want to let the rest of the system keep going. I'm not sure what happens to BadApp though (log, restart/kill, and obviously fix?).

That being said, I also understand the other point of view. BadApp is there for a reason and it cannot perform its function if it passes a `NULL` pointer and restarting it may just result in a restart loop.

Reply:
I think it's reasonable to document `kDif<ip>BadArg` as recoverable. It's up to the SW using the DIFs whether or not it uses that property but from a documentation point of view it's good to know what is and isn't possible.

Reply:
Ok, so it sounds like it's good to document the property "this return code means the device has been left in an improper (or unknown) state" vs "this return code means the operation was not performed, but/and the device is still in a usable state". 

It seems "Recoverable" isn't the right word for this, but that property is at least useful to document. I will try to work out a way of phrasing this.

Reply:
I've felt that all errors can be given one of a few "remediation strategies":
- Reboot the device. You're in an impossible situation.
- Retry with different arguments (in a lot of cases, this degenerates to "reboot the device").
- Retry with the same arguments. (E.g., you got quota'd or evicted.)

I think that `Error` and `BadArg` correspond to the first two, respectively, while you could imagine IPs which have errors with the third kind. I think a better phrasing might be along the lines of: "Every error must specifiy a remediation strategy, such as [...]." We should perhaps be careful as to what kinds of remediation strategies we allow, to avoid overly complicated caller-side control flow based on errors.